server/src/main/java/org/elasticsearch/ingest/IngestCtxMap.java

```diff
@@ -42,7 +42,12 @@ class IngestCtxMap extends CtxMap<IngestDocMetadata> {
         ZonedDateTime timestamp,
         Map<String, Object> source
     ) {
-        super(new HashMap<>(source), new IngestDocMetadata(index, id, version, routing, versionType, timestamp));
+        // if source is already a HashMap, then just use it as is, otherwise we'll need to copy into a new HashMap so that
+        // we have something mutable
+        super(
+            (source instanceof HashMap) ? source : new HashMap<>(source),
```

Comment:
I know right now we just have `HashMap` and `UnmodifiableMap` from `Map.of(...)` but we need to rethink this contract if we're going to specialize on `HashMap`.

The goal behind the check is for mutability, not necessarily for `HashMap`, it's just using that `instanceof` as a indicator of mutability.